Marlin/src/gcode/control/M3-M5.cpp

```diff
@@ -85,56 +85,57 @@ void GcodeSuite::M3_M4(const bool is_M4) {
   #if ENABLED(LASER_POWER_INLINE)
     if (parser.seen('I') == DISABLED(LASER_POWER_INLINE_INVERT)) {
       // Laser power in inline mode
+      planner.laser_inline.status.isInline = true;
+      planner.laser_inline.status.alwaysOn = false;
       cutter.inline_direction(is_M4); // Should always be unused
       #if ENABLED(SPINDLE_LASER_PWM)
         if (parser.seen('O')) {
           cutter.unitPower = cutter.power_to_range(parser.value_byte(), 0);
-          cutter.inline_ocr_power(cutter.unitPower); // The OCR is a value from 0 to 255 (uint8_t)
+          cutter.inline_power(cutter.unitPower); // The OCR is a value from 0 to 255 (uint8_t)
         }
         else
           cutter.inline_power(cutter.upower_to_ocr(get_s_power()));
       #else
-        cutter.set_inline_enabled(true);
+        cutter.inline_power(255);
+        cutter.unitpower = 255;
       #endif
       return;
     }
-    // Non-inline, standard case
-    cutter.inline_disable(); // Prevent future blocks re-setting the power
   #endif
-
-  planner.synchronize();   // Wait for previous movement commands (G0/G0/G2/G3) to complete before changing power
+  planner.synchronize();   // Wait for previous movement commands (G0/G1/G2/G3) to complete before changing power
   cutter.set_direction(is_M4);
 
   #if ENABLED(SPINDLE_LASER_PWM)
     if (parser.seenval('O')) {
       cutter.unitPower = cutter.power_to_range(parser.value_byte(), 0);
       cutter.set_ocr_power(cutter.unitPower); // The OCR is a value from 0 to 255 (uint8_t)
+      TERN_(planner.laser_inline.status.isInline, inline_ocr_power(cutter.unitPower))
```

Comment:
The first argument to `TERN_` has to be a configuration option, such as `LASER_POWER_INLINE`.